proselint/checks/__init__.py

```diff
@@ -1 +1,394 @@
+                start = i
+                s = 1
+                q = c
+            elif s == 1 and matching(c, q):
+                s = 2
+            elif s == 2:
+                if c in seps:
+                    ranges.append((start + 1, i - 1))
+                    start = None
+                    s = 0
+                else:
+                    s = 1
+            pc = c
+        return ranges
+
+    def position_in_ranges(ranges: list[tuple[int, int]], _position: int) -> bool:
+        return any(start <= _position < end for start, end in ranges)
+
+    return position_in_ranges(find_ranges(text), position)
+
+
+###############################################################################
+# The actual check-sub-functions used by the checks  ##########################
+###############################################################################
+
+
+def consistency_check(
+    text: str,
+    word_pairs: list,
+    err: str,
+    msg: str,
+    offset: int = 0,
+) -> list[ResultCheck]:
+    """Build a consistency checker for the given word_pairs."""
+    results = []
+
+    for w in word_pairs:
+        matches = [
+            list(re.finditer(w[0], text)),
+            list(re.finditer(w[1], text)),
+        ]
+
+        if len(matches[0]) > 0 and len(matches[1]) > 0:
+            idx_minority = len(matches[0]) > len(matches[1])
+
+            results += [
+                (
+                    m.start() + offset,
+                    m.end() + offset,
+                    err,
+                    msg.format(w[not idx_minority], m.group(0)),
+                    w[not idx_minority],
+                )
+                for m in matches[idx_minority]
+            ]
+
+    return results
+
+
+def preferred_forms_check(  # noqa: PLR0913, PLR0917
+    text: str,
+    items: list,
+    err: str,
+    msg: str,
+    ignore_case: bool = True,
+    offset: int = 0,
+) -> list[ResultCheck]:
+    """Build a checker that suggests the preferred form."""
+    flags = re.IGNORECASE if ignore_case else 0
+    regex = r"[\W^]{}[\W$]"  # TODO: replace with enum below?
+
+    return [
+        (
+            m.start() + 1 + offset,
+            m.end() + offset,
+            err,
+            msg.format(item[0], m.group(0).strip()),
+            item[0],
+        )
+        for item in items
+        for r in item[1]
+        for m in re.finditer(regex.format(r), text, flags=flags)
+    ]
+    # TODO: can we speed up str.format() ?
+    #       fast-string? or do padding already in checks
+
+
+# PADDINGS, add more if needed
+class Pd(str, Enum):
+    disabled = r"{}"
+    # choose for checks with custom regex
+    whitespace = r"\s{}\s"
+    # -> req whitespace around (no punctuation!)
+    sep_in_txt = r"(?:^|\W){}[\W$]"
+    # req non-text character around
+    # -> finds item as long it is surrounded by any non-word character:
+    #       - whitespace
+    #       - punctuation
+    #       - newline ...
```

Comment:
I would love some clarification as to what this is for, and if it's for paddings, why isn't it used in the function just prior?

Reply:
you are right, I've improved that further. the original code just had some magic hardcoded regex-paddings. how about:

```python
# Regex-PADDINGS
# - these can be handed to the check-functions
# - most efficient is `words_in_txt` but it does not work
#   for all use-cases
# - test with tools like https://regex101.com/ and optimize for low step-count
class Pd(str, Enum):
    disabled = r"{}"
    # choose for checks with custom regex

    safe_join = r"(?:{})"
    # can be filled with w1|w2|w3|...

    whitespace = r"\s{}\s"
    # req whitespace around (no punctuation!)

    sep_in_txt = r"(?:^|\W){}[\W$]"
    # finds item as long it is surrounded by any non-word character:
    # whitespace, punctuation, newline ...

    words_in_txt = r"\b{}\b"
    # much faster version of sep_in_txt, but more specialized, as
    # non A-z - characters at start & end of search-string don't match!
```

Reply:
I would be okay with that, but comments should come before entries, not after.